### PR TITLE
1309 bugfix

### PR DESCRIPTION
**Pull request by ScottKolmarWFI**
While regression models only have a single shapley value dataframe, classifier models produce a shapley value dataframe for every prediction class.
This branch contains logic to accommodate this new data structure.

Also fixed a bug which occurs when there are multiple joblib file types. If there are multiple joblib file types in the artifact directory the model artifact extraction method would pull the first one. Now, the logic iterates through each joblib file, and has an isinstance() type check which verifies that the joblib file produces an XGBModel class. This will need to be extended for other models object types; the optimal solution is probably to use the base sklearn model object class, which I believe is inherited by all sklearn and xgboost models.

### CODE REVIEW

File: src/sageworks/utils/extract_model_artifact.py [hunk 1]
```diff
@@ -84,21 +85,34 @@ def download_and_extract_model(model_artifact_uri):
...
+                model_object = joblib.load(model_file)
+
+                # Check type
+                if isinstance(model_object, XGBModel):
+                    print(f"{model_file} is a model object.")
+
+                    # Set return if type check passes
+                    model_return = model_object
+                else:
+                    print(f"{model_file} is NOT a model object.")
```
Comment (brifordwylie) [inline src/sageworks/utils/extract_model_artifact.py:106]: Trying to understand the new logic here, admittedly the old logic isn't great either :)

Looks like it loops around and basically picks the ^last^ thing it sees (that's a model) as the return. What use case has multiple models where we want the last one? Or I guess the other use case is we have stuff that ends in .joblib that aren't models?  Okay.. so thinking through this.. maybe the label encoder is getting joblib? So I guess this is good.. but means we're locked into XGBoost models for this to work... double shrug...

File: src/sageworks/artifacts/endpoints/endpoint.py [hunk 2]
```diff
@@ -427,9 +427,26 @@ def capture_performance_metrics(
...
+
+        # Single shap vals CSV for regressors
+        if model_type == ModelType.REGRESSOR.value:
+
+            # Format shap values into single dataframe
+            df_shap = pd.DataFrame(shap_vals, columns=X_pred.columns)
+
+            # Write shap vals to S3 Model Inference Folder
+            self.log.debug(f"Writing SHAP values to {self.model_inference_path}/inference_shap_values.csv")
+            wr.s3.to_csv(df_shap, f"{self.model_inference_path}/inference_shap_values.csv", index=False)
```
Comment (brifordwylie) [inline src/sageworks/artifacts/endpoints/endpoint.py:450]: It's fine but both endpoint.py and model.py are getting to be a LOT of lines of code. We need to think about pulling out some logic when/where we can. Shrug... perhaps most of `capture_performance_metrics` can be split out into some utility methods. I'll take that as an action item...

File: src/sageworks/artifacts/endpoints/endpoint.py [hunk 3]
```diff
@@ -452,7 +469,7 @@ def shap_values(model, X: pd.DataFrame) -> pd.DataFrame:
         # Note: For Tree-based models like decision trees, random forests, XGBoost, LightGBM,
         explainer = shap.TreeExplainer(model)
         shap_vals = explainer.shap_values(X)
-        return pd.DataFrame(shap_vals, columns=X.columns)
+        return shap_vals
```
Comment (brifordwylie) [inline src/sageworks/artifacts/endpoints/endpoint.py:472]: Update the return value to include the Union[] like you did in other places

**Review by brifordwylie — APPROVED** [review]
Okay.. read the comments, feel free to merge at your convenience. I'll change the Markdown logic so it doesn't try to parse the list of dataframes

File: src/sageworks/utils/extract_model_artifact.py (hunk 1 above)
Comment (ScottKolmarWFI) [reply src/sageworks/utils/extract_model_artifact.py:106]: Yeah in this particular case the first file in the model zip is a label encoder, which can't be used to calculate shapley values . So the idea is that that model zip may contain multiple objects as joblib files, but not all of them are "models". We need the actual XGBModel to extract the shapley values.

So all this block does is go through each joblib file, load it, and check the type. If the type is XGBModel, it will return that. If not, it will continue to iterate through the list.

I guess the return statement could actually be in the "if" block to make it cleaner, then it would just return the first XGBModel object. But while there are multiple joblib files in the model zip, there should only be 1 joblib that produces an XGBModel object. As far as I know.

Comment (ScottKolmarWFI) [reply src/sageworks/utils/extract_model_artifact.py:106]: To clarify, it doesn't always take the _last_ file. It takes the last file that is an XGBModel object. But that should only be one.

File: src/sageworks/artifacts/endpoints/endpoint.py (hunk 2 above)
Comment (ScottKolmarWFI) [reply src/sageworks/artifacts/endpoints/endpoint.py:450]: The issue I ran into with writing some of those methods as separate classes is that I ended up making a circular import from the Model class, or whatever artifact class I wanted to make methods for.

I generally don't feel like it's an issue to have a lot of lines of code in a single file when it pertains to a single class, but maybe there's a higher level reasoning to that.

File: src/sageworks/artifacts/endpoints/endpoint.py (hunk 3 above)
Comment (ScottKolmarWFI) [reply src/sageworks/artifacts/endpoints/endpoint.py:472]: So this now actually returns an np.array - I'll make the change.

The np.array here ends up being (# of prediction classes x # samples x # features), which later gets split into a separate dataframe for each prediction class, so that's when the Union of the list[pd.DataFrame] type comes in.
